### ISSUE TITLE
StoreWatcherMixin Widget not rebuilding 

### ISSUE THREAD

**Issue by JamesWatton** [google/flutter_flux#34]
Hi,

I've been wrestling with this for a while now and can't see what is going wrong, but my StoreWatcherMixin widget does not get build() called after setting state in the flux store.

Main.dart:
```
void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  State createState() => new AppState();
}

class AppState extends State<MyApp>
  with StoreWatcherMixin<MyApp> {

  var appStore;

  @override
  void initState() {
    super.initState();
    appStore = listenToStore(_appStoreToken);
  }

  @override
  Widget build(BuildContext context) {
      ...
      return new SecondPage(_appStoreToken);
   }
```

The store:
```
final Action<List<Foo>> setFoos = new Action<List<Foo>>();
final AppStore _store = new AppStore();
final StoreToken _appStoreToken = new StoreToken(_store);

class AppStore extends Store {
  List<Foo> _allFoos = new List<Foo>();
  List<ArkhamDBCardPack> get allFoos => _allFoos;

  AppStore() {
    triggerOnAction(setFoos, (value) => _allFoos = value);
  }
}
```

Second page:
```
class SecondPage extends StatefulWidget {

  final StoreToken _storeToken;
  SecondPage(this._storeToken);

  @override
  State createState() => new SecondPageState(_storeToken);
}

class SecondPageState extends State<SecondPage>
  with StoreWatcherMixin<SecondPage> {

  var appStore;
  final StoreToken _storeToken;
  List<Foo> foos;

  SecondPageState(this._storeToken);

  @override
  void initState() {
    super.initState();
    debugPrint('x');
    appStore = listenToStore(_storeToken);
    foos = appStore.allFoos;
  }

  @override
  Widget build(BuildContext context) {
    debugPrint('rebuilding');
    return new IconButton(
              icon: new Icon(Icons.refresh),
              onPressed: () async {
                var data = await fetchFoos();
                setFoos.call(data);
              },
            )
   }
}
```

This isn't the exact code, but its close enough. The first time it runs, I press the button on the second page to fetch the data and call setFoos. debugPrints show that it has set the value correctly into the store, but the build() and debugPrint('rebuilding') are never called a second time.

What am I doing wrong? I really want to get this working as flutter and flux seem like a great tech. Thanks in advance!

**Comment by jimbeveridge** [comment]
You didn't show your "fetchFoos" function, but I'm wondering if it ever returns. Set a breakpoint at "setFoos.call(data)" to make sure fetchFoos() return and then and then step into the call() function.  You should end up in StoreWatcherState._handleStoreChanged(), which should then call setState(). After that, reaching your build() function should be straightforward.

If running the debugger is problematic, try putting some debugPrint statements in onPressed and in StoreWatcherState._handleStoreChanged.

BTW, as a matter of convenience, the "call" function is special. To call such a function, you can just say:
setFoos(data);

**Comment by JamesWatton** [comment]
Thanks for the reply. "fetchFoos" does return a list as expected, I can attach a breakpoint and see that, but I am unable to step into setFoos. Hovering over it with the mouse cursor inspects it and shows that setFoos is an Action with a _listeneres list of 0 items - it should have one (AppStore)?

edit - Okay, so when I called setFoos from within main.dart (which also had the store within the same dart file) it works fine. When I call setFoos from secondpage.dart file, it would not have any listeners. I moved the store out into its own file and now it works from everywhere. I must have messed up with the scopes of my variables but its working now! Thanks for the guidance!
